src/pybind/mgr/dashboard/controllers/rgw.py

```diff
@@ -84,40 +83,169 @@ def get(self, svc_id):
         return daemon
 
 
-@ApiController('/rgw/proxy')
-@AuthRequired()
-class RgwProxy(BaseController):
+class RgwRESTController(RESTController):
 
-    @Proxy()
-    def __call__(self, path, **params):
+    def proxy(self, method, path, params=None, json_response=True):
         try:
             rgw_client = RgwClient.admin_instance()
-
-            method = cherrypy.request.method
-            data = None
-
-            if cherrypy.request.body.length:
-                data = cherrypy.request.body.read()
-
-            return rgw_client.proxy(method, path, params, data)
-        except RequestException as e:
-            # Always use status code 500 and NOT the status that may delivered
-            # by the exception. That's because we do not want to forward e.g.
-            # 401 or 404 that may trigger unwanted actions in the UI.
-            cherrypy.response.headers['Content-Type'] = 'application/json'
-            cherrypy.response.status = 500
-            return json.dumps({'detail': str(e)}).encode('utf-8')
+            result = rgw_client.proxy(method, path, params, None)
+            if json_response and result != '':
+                result = json.loads(result.decode('utf-8'))
+            return result
+        except (DashboardException, RequestException) as e:
+            raise DashboardException(e, http_status_code=500, component='rgw')
 
 
 @ApiController('/rgw/bucket')
 @AuthRequired()
-class RgwBucket(RESTController):
+class RgwBucket(RgwRESTController):
+
+    def list(self):
+        return self.proxy('GET', 'bucket')
+
+    def get(self, bucket):
+        return self.proxy('GET', 'bucket', {'bucket': bucket})
 
     def create(self, bucket, uid):
         try:
             rgw_client = RgwClient.instance(uid)
             return rgw_client.create_bucket(bucket)
         except RequestException as e:
-            cherrypy.response.headers['Content-Type'] = 'application/json'
-            cherrypy.response.status = 500
-            return {'detail': str(e)}
+            raise DashboardException(e, http_status_code=500, component='rgw')
+
+    def set(self, bucket, bucket_id, uid):
+        return self.proxy('PUT', 'bucket', {
+            'bucket': bucket,
+            'bucket-id': bucket_id,
+            'uid': uid
+        }, json_response=False)
+
+    def delete(self, bucket, purge_objects='true'):
+        return self.proxy('DELETE', 'bucket', {
+            'bucket': bucket,
+            'purge-objects': purge_objects
+        }, json_response=False)
+
+
+@ApiController('/rgw/user')
+@AuthRequired()
+class RgwUser(RgwRESTController):
+
+    def list(self):
+        return self.proxy('GET', 'metadata/user')
+
+    def get(self, uid):
+        return self.proxy('GET', 'user', {'uid': uid})
+
+    def create(self, uid, display_name, email=None, max_buckets=None,
+               suspended=None, generate_key=None, access_key=None,
+               secret_key=None):
```

Comment:
I'd be in favor of changing the signature to `def create(self, uid, display_name, **kwargs):` and then just forwarding `kwargs` to the proxy method.

Reply:
That's not possible, the RGW Admin Ops API uses parameters names like ``generate-key`` or ``access-key``. So we need to rename them e.g. from ``access_key```=> ``access-key`` because Python does not support hyphens in parameters.